marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -0,0 +1,154 @@
+module Marlowe.Linter
+  ( lint
+  , State(..)
+  , Position
+  , _holes
+  , _negativeDeposits
+  , _negativePayments
+  ) where
+
+import Prelude
+import Data.Array (catMaybes, fold, foldMap, (:))
+import Data.BigInteger (BigInteger)
+import Data.Lens (Lens', over, set, view)
+import Data.Lens.Iso.Newtype (_Newtype)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (class Newtype)
+import Data.Symbol (SProxy(..))
+import Data.Tuple.Nested (type (/\), (/\))
+import Marlowe.Holes (Action(..), Case(..), Contract(..), Holes, Term(..), Value(..), getHoles, insertHole)
+import Text.Parsing.StringParser (Pos)
+
+type Position
+  = { start :: Pos, end :: Pos }
+
+newtype State
+  = State
+  { holes :: Holes
+  , negativePayments :: Array Position
+  , negativeDeposits :: Array Position
+  }
+
+derive instance newtypeState :: Newtype State _
+
+instance semigroupState :: Semigroup State where
```

Comment:
I could be wrong here, but I _think_ purescript records are semigroups if their fields are semigroups. So you might be able to derive a newtype instance here. Ditto for the monoid.

Could be wrong, but it would be sweet if it works.

Reply:
nice